File: List and Node Manipulation (C)/List.c
// Student: Josh, Tsung Chi, Shih
// CruzID/CruzID#: jtshih/1436140
// Class: CMPS 101
// Date: 1/23/2015
// Filename: List.c


#include<stdio.h>
#include<stdlib.h>
#include "List.h"

//structs

//private NodeObj type
typedef struct NodeObj{
	int data;
	struct NodeObj* next;
	struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj{
   Node front;
   Node back;
   Node cursor;
   int cursorIndex;
} ListObj;

//constructors-destructors

// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   N->prev = NULL;
   return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

//newList()
//Returns reference to the new empty List object.
List newList(void){
	List L;
	L = malloc(sizeof(ListObj));
	L->front = L->back = L->cursor = NULL;
	return(L);
}

// freeList()
// Frees all heap memory associated with List *pL, and sets *pl to NULL.S
void freeList(List* pL){
   if(pL!=NULL && *pL!=NULL) { 
      while(length(*pL) != 0) { 
         deleteFront(*pL); 
      }
      free(*pL);
      *pL = NULL;
   }
}

//Access functions -----------------------------------------------------------
int length(List L){
   if( L->front ==NULL ){
      return 0;
   }
   int i=0;
   Node walker = L->front;
   i = 1;
   while(walker != L->back)
   {
      walker = walker->next;
      i++;
   }
   return i;
}

int getIndex(List L){
   if(L->cursor == NULL)
      return -1;
   else{
      Node N = L->front;
      int i = 0;
      while(N != L->cursor)
      {
         i++;
	 N = N->next;
      }
      return i;
   }
}

int front(List L){
   if(length(L) < 0){
      printf("front() error! Empty List!\n");
      exit(1);
   }
   return L->front->data;
}

int back(List L){
   if(length(L) < 0){
      printf("back() error! Empty List!\n");
      exit(1);
   }
   return L->back->data;
}

int getElement(List L){
   if(length(L) < 0){
      printf("getElement() error! Empty List!\n");
      exit(1);
   }
   if(L->cursor == NULL){
      printf("getElement() error! Cursor isn't on the list!\n");
      exit(1);
   }
   return L->cursor->data;
}

int equals(List A, List B){
   Node aWalker = A->front;
   Node bWalker = B->front;
   while(aWalker != NULL && bWalker != NULL){
      if(aWalker->data != bWalker->data)
         return 0;
      aWalker = aWalker->next;
      bWalker = bWalker->next;
   }
   if(bWalker != NULL)
      return 0;
   if(aWalker != NULL)
      return 0;
   return 1;
}

//Manipulation functions------------------------------------------------------
void clear(List L){
   Node N = NULL;
   while(N!=L->back){
      deleteFront(L);
      N = L->front;
   }
   if(N==L->back){
      L->front = L->back = L->cursor = NULL;
      freeNode(&N);
   }
   else{
      printf("List Error: calling clear()\n");
      exit(1);
   }
}
void moveTo(List L, int i){
   if(i < 0 || i >= length(L))
	L->cursor = NULL;
   else{
      Node N = L->front;
      for(int j = 0; j<i; j++){
         N = N->next;
      }
      L->cursor = N;
   }
}
void movePrev(List L){
   if(getIndex(L) == -1)
      return;
   moveTo(L, getIndex(L)-1);
}
void moveNext(List L){
   if(getIndex(L) == -1)
      return;
   moveTo(L, getIndex(L)+1);
}
void prepend(List L, int data){
   Node N = newNode(data);
   if (length(L) == 0){
      L->front = L->back = N;
   }
   else{
      N->next = L->front;
      L->front->prev = N;
      L->front = N;
   }
}
void append(List L, int data){
   Node N = newNode(data);
   
   if (length(L) == 0){
      L->front = L->back = N;
   }
   else{
      L->back->next = N;
      N->prev = L->back;
      L->back = N;
   }
}
void insertBefore(List L, int data){
   if(getIndex(L)==-1){
      printf("Cursor does not exist! Can't insertBefore()\n");
      exit(1);
   }
   if(length(L)<=0){
      printf("List does not exist! Can't insertBefore()\n");
      exit(1);
   }
   if(L->cursor == L->front){
      prepend(L, data);
      return;
   }
   Node N = newNode(data);
   Node before = L->cursor->prev;
   before->next = N;
   N->prev = before;
   N->next = L->cursor;
   L->cursor->prev = N;
}
void insertAfter(List L, int data){
   if(getIndex(L)==-1){
      printf("Cursor does not exist! Can't insertAfter()\n");
      exit(1);
   }
   if(length(L)<=0){
      printf("List does not exist! Can't insertAfter()\n");
      exit(1);
   }
   if(L->cursor == L->back){
      append(L, data);
      return;
   }
   Node N = newNode(data);
   Node after = L->cursor->next;
   L->cursor->next = N;
   N->prev = L->cursor;
   N->next = after;
   after->prev = N;
}
void deleteFront(List L){
   Node N = NULL;
   if( length(L) <= 0 ){
      printf("Empty List deleteFront() Error!\n");
      exit(1);
   }
   N = L->front;
   if(L->cursor == N)
      L->cursor = NULL;
   if( length(L)>1 ) {
      L->front = L->front->next; 
      L->front->prev = NULL;
   }
   else { 
      L->front = L->back = NULL;
   }

   freeNode(&N);
}
void deleteBack(List L){
   Node N = NULL;
   if( length(L) <= 0 ){
      printf("Empty List deleteFront() Error!\n");
      exit(1);
   }
   N = L->back;
   if(L->cursor == N)
      L->cursor = NULL;
   if(length(L)>1){
      L->back = L->back->prev;
      L->back->next = NULL;
   }
   else{
      L->back = L->front = NULL;
   }
   freeNode(&N);
}
void delete(List L){
   if(getIndex(L)==-1){
      printf("Cursor does not exist! Can't delete()\n");
      exit(1);
   }
   if(length(L)<=0){
      printf("List does not exist! Can't delete()\n");
      exit(1);
   }
   Node before = L->cursor->prev;
   before->next = L->cursor->next;
   L->cursor->next->prev = before;
   L->cursor->next = L->cursor->prev = NULL;
   Node N = L->cursor;
   freeNode(&N);
   L->cursor = NULL;   
}
//Other operations -----------------------------------------------------------
void printList(FILE* out, List L){
   Node N = NULL;
   for(N = L->front; N != NULL; N = N->next){
//      if(N == L->cursor)
//         printf(">");
      printf("%d ", N->data);
   }
}
List copyList(List L){
   List T = newList();
   Node N = L->front;
   for(int i=0; i<length(L); i++){
      append(T, N->data);
      N = N->next;
   }
   return T;
}
List concatList(List A, List B){
   List T = newList();
   Node N = A->front;
   Node Nb = B->front;
   for(int i=0; i<length(A)+length(B); i++){
      if(i<length(A)){
         append(T, N->data);
	 N = N->next;
      }
      else{
         append(T, Nb->data);
	 Nb = Nb->next;
      }
   }
   return T;
}

File: List and Node Manipulation (Java)/List.java
// Student: Josh, Tsung Chi, Shih
// CruzID/CruzID#: jtshih/1436140
// Class: CMPS 101
// Date: 1/14/2015
// Filename: List.java
// Description: List file that has private nodes. ADT: doubly linked list.

public class List {
	
	//CURSOR TOGGLE
    private boolean cursorToggle = true;

	private class Node {
        private int data;
        private Node next;
        private Node prev;
        Node(int data)
        {
            this.data = data;
            next = null;
            prev = null;
        }
        private Node()
        {
        	next = null;
        	prev = null;
        }
		public String toString() 
        { 
        	return String.valueOf(data); 
        }
    }
    private Node root = null;
    private Node end = null;
    private Node cursor;
    private int cursorIndex;
    //CONSTRUCTOR********************************************************************************
    List()
    {
    	cursor = new Node();
    	cursorIndex = -1;
    }
    //ACCESS FUNCTIONS***************************************************************************
    int length()
    {
    	int i=0;
    	Node walker = root;
    	if(root == end) 						//fix1
    		return 0;
    	while(walker!=null)
    	{
    		walker = walker.next;
    		i++;
    	}
    	return i;
    }
    int getIndex() //cursor function
    {
    	return cursorIndex;
    }
    int front()
    {
    	if(root == null)
    		throw new RuntimeException("Precondition Not Met");
    	return root.data;
    }
    int back()
    {
    	if(end==null)
    		throw new RuntimeException("Precondition Not Met");
    	return end.data;
    }
    int getElement()
    {
       	if(length()>0 && getIndex() >= 0)
       		return cursor.data;
       	else
       		throw new RuntimeException("Precondition Not Met");
    }
    boolean equals(List L)
    {
    	Node listWalker = root;
    	Node list2Walker = L.root;
    	
    	if(length() != L.length())
    		return false;
    	for(int i=0; i<length(); i++)
    	{
    		if(listWalker.data != list2Walker.data)
    			return false;
    			
    		listWalker = listWalker.next;
    		list2Walker = list2Walker.next;
    	}
    	return true;
    }
    //MANIPULATION FUNCTIONS**********************************************************************
    void clear()
    {
    	root = null;
    	end = null;
    }
    void moveTo(int i)
    {
    	if(root == null)
    		throw new RuntimeException("Precondition Not Met");
//    	System.out.println("bloop");
    	if(i>=0 || i<=length()-1)
    	{
    		Node walker = root;
	    	for(int x=0; x<i; x++)
	    		walker = walker.next;
	    	cursor = walker;
	    	cursorIndex = i;
//	    	System.out.println(": "+i);
    	}
    }
    void movePrev()
    {
    	if(cursor==root)
    	{
    		Node blank = new Node(-1);
    		cursor = blank;
    		cursorIndex = -1;
    		return;
    	}
    	cursor = cursor.prev;
    	cursorIndex--;
    }
    void moveNext()
    {
    	if(cursor==end)
    	{
    		Node blank = new Node(-1);
    		cursor = blank;
    		cursorIndex = -1;
    		return;
    	}
    	cursor = cursor.next;
    	cursorIndex++;
    }
    void prepend(int a)
    {
    	Node temp = new Node(a);
    	if(root==null)
    	{
    		end = temp;
    	}
    	else
    	{
//    		System.out.println("BKAHDKAHFSDFJSF");
    		temp.next = root;
	    	root.prev = temp;
    	}
    	root = temp;
    }
    void append(int a)
    {
		Node temp = new Node(a);
    	if(root == null)//if list is empty
    		root = temp;
    	else
    	{
			temp.prev = end;
			end.next = temp;
    	}
		end = temp;
    }
    void insertBefore(int a)
    {
    	if(length()<0)
    		throw new RuntimeException("Precondition Not Met");
    	if(getIndex() < 0)
    		throw new RuntimeException("Precondition Not Met");
    	if(cursor==root)
    		prepend(a);
    	else
    	{
    		Node walker = cursor;
	    	Node temp = new Node(a);
			temp.next = walker;
			
			if(cursor.prev == null)
				root = temp;
			else
				cursor.prev.next = temp;
			walker.prev = temp;
    	}
    	

		cursorIndex++;
    }
    void insertAfter(int a)
    {
    	if(length()<0)
    		throw new RuntimeException("Precondition Not Met");
    	if(getIndex() < 0)
    		throw new RuntimeException("Precondition Not Met");
    	
    	Node walker = cursor;
    	walker=walker.next;
    	
    	if(walker != end)
    	{
    		Node temp = new Node(a);
    		cursor.next = temp;
    		temp.prev = cursor;
    		walker.prev = temp;
    		temp.next = walker;
    	}
    	else
    	{
    		append(a);
    	}

    }
    void deleteFront()
    {
    	if(length()<0)
    		throw new RuntimeException("Precondition Not Met");
    	Node temp = root;
    	root = root.next;
    	root.prev = null;
    	temp.next = null;
    	temp = root;
//    	for(int i=0; i<length(); i++)
//    	{
//    		temp.index = i;
//    		temp = temp.next;
//    	}
    }
    void deleteBack()
    {
    	if(length()<0)
    		throw new RuntimeException("Precondition Not Met");
    	Node temp = end.prev;
    	temp.next = null;
    	end.prev = null;
    	end = temp;
    }
    void delete()
    {
    	Node walker = cursor;
    	walker = walker.next;
    	if(length()<0)
    		throw new RuntimeException("Precondition Not Met");
    	if(getIndex()<0)
    		throw new RuntimeException("Precondition Not Met");
    	
    	if(cursor == root)
    		deleteFront();
    	else if(cursor == end)
    		deleteBack();
    	else
    	{
    		System.out.println("hello");
	    	Node before = cursor.prev;
	    	before.next = walker;
	    	walker.prev = before;
	    	cursor.next = null;
	    	cursor.prev = null;
    	}
    }
    //OTHER METHODS ***************************************************
    public String toString()
    {
    	Node walker = root;
    	String spit = "";
    	spit += walker.data;
    	walker = walker.next;
		if(cursor == root && cursorToggle == true)
			spit += "<";
    	while(walker != null)
    	{
    		spit +=" "+walker.data;
    		if(walker == cursor && cursorToggle == true) 
    			spit += "<";

    		walker = walker.next;
    	}
    	return spit;
    }
    List copy()
    {
    	List bloop = new List();
    	Node walker = root;
    	while(walker != null)
    	{
    		bloop.append(walker.data);
    		walker = walker.next;
    	}
    	return bloop;
    }
    List concat(List L)
    {
    	List bloop = new List();
    	bloop = copy();
    	bloop.end.next = L.root;
    	L.root.prev = bloop.end;
    	return bloop;
    }
}
